docs/src/main/asciidoc/security-openid-connect-client-registration.adoc

```diff
@@ -0,0 +1,332 @@
+        }
+        return null;
+    }
+
+    // Convert metadata of registered clients to OidcTenantConfig
+    private OidcTenantConfig createTenantConfig(String tenantId, RegisteredClient client) {
+        ClientMetadata metadata = client.getMetadata();
+
+        OidcTenantConfig oidcConfig = new OidcTenantConfig();
+        oidcConfig.setTenantId(tenantId);
+        oidcConfig.setAuthServerUrl(authServerUrl);
+        oidcConfig.setApplicationType(ApplicationType.WEB_APP);
+        oidcConfig.setClientName(metadata.getClientName());
+        oidcConfig.setClientId(metadata.getClientId());
+        oidcConfig.getCredentials().setSecret(metadata.getClientSecret());
+        String redirectUri = metadata.getRedirectUris().get(0);
+        oidcConfig.getAuthentication().setRedirectPath(URI.create(redirectUri).getPath());
+        return oidcConfig;
+    }
+}
+----
+
+=== Register clients on demand
+
+You can register new clients on demand. 
+You can add new clients to the existing, already configured `OidcClientConfiguration` or to a newly created `OidcClientConfiguration`.
+
+Start from configuring one or more OIDC client registrations:
+
+[source,properties]
+----
+quarkus.oidc-client-registration.auth-server-url=${quarkus.oidc.auth-server-url}
+----
+
+The above configuration is sufficient for registering new clients using this configuration. For example:
+
+[source,java]
+----
+package io.quarkus.it.keycloak;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import io.quarkus.oidc.OidcRequestContext;
+import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.oidc.OidcTenantConfig.ApplicationType;
+import io.quarkus.oidc.TenantConfigResolver;
+import io.quarkus.oidc.client.registration.ClientMetadata;
+import io.quarkus.oidc.client.registration.OidcClientRegistration;
+import io.quarkus.oidc.client.registration.OidcClientRegistrationConfig;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+@Singleton
+public class CustomTenantConfigResolver implements TenantConfigResolver {
+
+    @Inject
+    OidcClientRegistration clientReg;
+
+    @Inject
+    @ConfigProperty(name = "quarkus.oidc.auth-server-url")
+    String authServerUrl;
+
+
+    @Override
+    public Uni<OidcTenantConfig> resolve(RoutingContext routingContext,
+            OidcRequestContext<OidcTenantConfig> requestContext) {
+        if (routingContext.request().path().endsWith("/protected/oidc-client-reg-existing-config")) {
+            // New client registration done dynamically at the request time using the configured client registration
+            
+            ClientMetadata metadata = createMetadata("http://localhost:8081/protected/dynamic-tenant",
+                    "Dynamic Tenant Client");
+
+            return clientReg.registerClient(metadata).onItem().transform(r -> 
```

Comment:
Do you have plans to somehow store the registration access token so that the client can update itself in the future? I don't know how it would work but if the Quarkus application is using some storage (e.g.: JPA) and is able to store the token.

Reply:
@pedroigor It is already supported :-).

`OidcClientRegistration` gives `RegisteredClient` representing the already registered client which keeps that cient specific registration token. One of the test `TenantConfigResolver` uses it to [re-read](https://github.com/quarkusio/quarkus/pull/41866/files#diff-6154f623d750cd5f9dde30f4e8c7cbce822efa06ddd9aaf6e54228eb96a7a204R55) the given `RegisteredClient`'s metadata, update did not work for Keycloak, well it does work, but there is a strange side-effect impacting ID tokens (and the good news, it was already fixed in Keycloak, https://github.com/keycloak/keycloak/issues/31062, so I'll try the update a bit later one 25.0.3 is out).
But yeah, full CRUD is supported

Reply:
I haven't looked yet into persisting these client specific registration tokens though, hmm, as indeed, after the restart, it will be gone. 
I think it will be up to users how to persist them, but we do need to let users recreate already registered `RegisteredClient`, I guess, to start with, I'll add a few builder methods to create a `RegisteredClient` instance from the existing `ClientMetadata` and the reg token...

Reply:
Just an update here, a doc section has been added suggesting a strategy for avoiding duplicate registrations, `RegisteredClient` has had 2 methods added for users be able to persist the client specific reg URL and token.